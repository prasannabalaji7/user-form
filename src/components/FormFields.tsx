import React, { ChangeEvent } from "react";
import { Form, Row, Col } from "react-bootstrap";

export interface FormFieldProps {
	fieldId: string;
	testId: string;
	label: string;
	isValidClass: string;
	isReadOnly: boolean;
	fieldValue: string;
	errorMsg: string;
	fieldValid: boolean;
	userAction: (e: ChangeEvent<HTMLInputElement>) => void;
}

export const FormField: React.FC<FormFieldProps> = (props) => {
	const {
		label,
		isValidClass,
		isReadOnly,
		fieldValue,
		errorMsg,
		userAction,
		fieldId,
		testId,
		fieldValid,
	} = props;
	return (
		<Form.Group as={Row} id={fieldId}>
			<Form.Label column sm={4}>
				{label}
			</Form.Label>
			<Col sm={8}>
				<Form.Control
					className={isValidClass}
					readOnly={isReadOnly}
					type="text"
					placeholder="Name"
					data-testId={testId}
					value={fieldValue}
					onChange={userAction}
					required
				/>
				{!fieldValid && (
					<Form.Control.Feedback className="d-block" type="invalid">
						{errorMsg}
					</Form.Control.Feedback>
				)}
			</Col>
		</Form.Group>
	);
};