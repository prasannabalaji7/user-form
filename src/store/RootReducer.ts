import { Action, Reducer } from 'redux';

export interface ProfileData {
    userProfileName: string;
    userProfileRole: string;
    userProfileCountry: string;
}

export interface UserFormData {
    userName: string;
    userEmail: string;
    userRole: string;
    userMobile: string;
    userCountry: string;
}

export interface InitialStateInterface {
    profileData: ProfileData;
    userFormData: UserFormData;
    formValid: boolean;
    isEditBtnVisible: boolean;
}

export const initialState: InitialStateInterface = {
    userFormData: {
        userName: 'User Name',
        userEmail: 'email@domain.com',
        userRole: 'User Role',
        userMobile: '+9311111111',
        userCountry: 'Afghanistan',
    },
    profileData: {
        userProfileName: 'User Name',
        userProfileRole: 'User Role',
        userProfileCountry: 'Afghanistan',
    },
    formValid: true,
    isEditBtnVisible: false,
};

export interface DispatchAction extends Action {
    payload: any;
}

export const RootReducer: Reducer<InitialStateInterface, DispatchAction> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case 'nameChange':
            return {
                ...state,
                userFormData: {
                    ...state.userFormData,
                    userName: action.payload.userName,
                },
            };
        case 'emailChange':
            return {
                ...state,
                userFormData: {
                    ...state.userFormData,
                    userEmail: action.payload.userEmail,
                },
            };
        case 'roleChange':
            return {
                ...state,
                userFormData: {
                    ...state.userFormData,
                    userRole: action.payload.userRole,
                },
            };
        case 'numberChange':
            return {
                ...state,
                userFormData: {
                    ...state.userFormData,
                    userMobile: action.payload.userMobile,
                },
            };
        case 'countryChange':
            return {
                ...state,
                userFormData: {
                    ...state.userFormData,
                    userCountry: action.payload.userCountry,
                },
            };
        case 'validateSubmit':
            return { ...state, formValid: action.payload.formValid };
        case 'onEdit':
            return {
                ...state,
                isEditBtnVisible: !action.payload.isEditBtnVisible,
            };
        case 'onSubmit':
            return {
                ...state,
                profileData: {
                    userProfileName: action.payload.userName,
                    userProfileRole: action.payload.userRole,
                    userProfileCountry: action.payload.userCountry,
                },
                isEditBtnVisible: !action.payload.isEditBtnVisible,
            };
        case 'onCancel':
            return { ...initialState };
        default:
            return { ...state };
    }
};