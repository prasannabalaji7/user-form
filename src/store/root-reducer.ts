import { Action, Reducer } from 'redux';

export interface ProfileData {
    userProfileName: string;
    userProfileRole: string;
    userProfileCountry: string;
    file: any;
}

export interface InitialStateInterface {
    userName: string;
    email: string;
    role: string;
    mobile: string;
    country: string;
    profileData: ProfileData;
    formValid: boolean;
    isEditEnabled: boolean;
}

export const initialState: InitialStateInterface = {
    userName: 'User Name',
    email: 'email@domain.com',
    role: 'User Role',
    mobile: '+9311111111',
    country: 'Afghanistan',
    isEditEnabled: false,
    profileData: {
        userProfileName: 'User Name',
        userProfileRole: 'User Role',
        userProfileCountry: 'Afghanistan',
        file: '',
    },
    formValid: true,
};

export interface DispatchAction extends Action {
    payload: any;
}

export const rootReducer: Reducer<InitialStateInterface, DispatchAction> = (
    state = initialState,
    action
) => {
    switch (action.type) {
        case 'nameChange':
            return { ...state, userName: action.payload.userName };
        case 'emailChange':
            return { ...state, email: action.payload.email };
        case 'roleChange':
            return { ...state, role: action.payload.role };
        case 'numberChange':
            return { ...state, mobile: action.payload.mobile };
        case 'countryChange':
            return { ...state, country: action.payload.country };
        case 'validateSubmit':
            return { ...state, formValid: !action.payload.formValid };
        case 'onEdit':
            return { ...state, isEditEnabled: !action.payload.isEditEnabled };
        case 'onSubmit':
            return {
                ...state,
                profileData: {
                    userProfileName: action.payload.userName,
                    userProfileRole: action.payload.role,
                    userProfileCountry: action.payload.country,
                    file: '',
                },
                isEditEnabled: !action.payload.isEditEnabled,
            };
        case 'onCancel':
            return { ...initialState };
        default:
            return { ...state };
    }
};