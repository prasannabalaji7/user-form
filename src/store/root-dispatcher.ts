import { Dispatch } from 'redux';
import { DispatchAction } from './root-reducer';

export class RootDispatcher {
    private readonly dispatch: Dispatch<DispatchAction>;

    constructor(dispatch: Dispatch<DispatchAction>) {
        this.dispatch = dispatch;
    }

    nameChange = (userName: string) =>
        this.dispatch({ type: 'nameChange', payload: { userName } });

    emailChange = (userEmail: string) =>
        this.dispatch({ type: 'emailChange', payload: { userEmail } });

    roleChange = (userRole: string) =>
        this.dispatch({ type: 'roleChange', payload: { userRole } });

    numberChange = (userMobile: string) =>
        this.dispatch({ type: 'numberChange', payload: { userMobile } });

    countryChange = (userCountry: string) =>
        this.dispatch({ type: 'countryChange', payload: { userCountry } });

    validateSubmit = (formValid: boolean) =>
        this.dispatch({ type: 'validateSubmit', payload: { formValid } });

    onCancel = () => this.dispatch({ type: 'onCancel', payload: {} });

    onEdit = (isEditBtnVisible: boolean) =>
        this.dispatch({ type: 'onEdit', payload: { isEditBtnVisible } });

    onSubmit = (
        userName: string,
        userEmail: string,
        userRole: string,
        userMobile: string,
        userCountry: string,
        isEditBtnVisible: boolean
    ) =>
        this.dispatch({
            type: 'onSubmit',
            payload: {
                userName,
                userEmail,
                userRole,
                userMobile,
                userCountry,
                isEditBtnVisible,
            },
        });
}

export default RootDispatcher;