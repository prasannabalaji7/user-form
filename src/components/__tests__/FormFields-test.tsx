import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";
import { FormField } from "../FormFields";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { initialState } from '../../store/RootReducer';

test("what dom renderer is inital load will remain after theere is change in state", () => {
    const mockStore = configureStore();
    const store = mockStore(initialState);
    const tree = renderer
        .create(
            <Provider store={store}>
                <FormField
                    fieldId="anyId"
                    testId="anyField"
                    label="Role"
                    isReadOnly={false}
                    fieldValue="Test 2"
                    errorMsg="errorMsg"
                    userAction={() => {}}
                    type="text"
                    validate={(text: string, country?: string) => false}
                />
            </Provider>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});