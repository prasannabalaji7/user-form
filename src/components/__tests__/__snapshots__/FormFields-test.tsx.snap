// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`what dom renderer is inital load will remain after theere is change in state 1`] = `
<div
  className="form-group row"
  id="anyId"
>
  <label
    className="form-label col-form-label col-sm-4"
  >
    Role
  </label>
  <div
    className="col-sm-8"
  >
    <input
      className="form-control"
      data-testId="anyField"
      onChange={[Function]}
      readOnly={false}
      required={true}
      type="text"
      value="Test 2"
    />
    <div
      className="d-block invalid-feedback"
    >
      errorMsg
    </div>
  </div>
</div>
`;
